fdbserver/storageserver.actor.cpp

```diff
@@ -1966,21 +2059,29 @@ ACTOR Future<Void> fetchKeys( StorageServer *data, AddingShard* shard ) {
 				if (this_block.more) {
 					Key nfk = this_block.readThrough.present() ? this_block.readThrough.get() : keyAfter( this_block.end()[-1].key );
 					if (nfk != keys.end) {
-						std::deque< Standalone<VerUpdateRef> > updatesToSplit = std::move( shard->updates );
+						state std::deque< Standalone<VerUpdateRef> > updatesToSplit = std::move( shard->updates );
 
 						// This actor finishes committing the keys [keys.begin,nfk) that we already fetched.
 						// The remaining unfetched keys [nfk,keys.end) will become a separate AddingShard with its own fetchKeys.
 						shard->server->addShard( ShardInfo::addingSplitLeft( KeyRangeRef(keys.begin, nfk), shard ) );
 						shard->server->addShard( ShardInfo::newAdding( data, KeyRangeRef(nfk, keys.end) ) );
 						shard = data->shards.rangeContaining( keys.begin ).value()->adding;
-						auto otherShard = data->shards.rangeContaining( nfk ).value()->adding;
+						state AddingShard* otherShard = data->shards.rangeContaining( nfk ).value()->adding;
 						keys = shard->keys;
 
 						// Split our prior updates.  The ones that apply to our new, restricted key range will go back into shard->updates,
 						// and the ones delivered to the new shard will be discarded because it is in WaitPrevious phase (hasn't chosen a fetchVersion yet).
 						// What we are doing here is expensive and could get more expensive if we started having many more blocks per shard. May need optimization in the future.
-						for(auto u = updatesToSplit.begin(); u != updatesToSplit.end(); ++u)
-							splitMutations( data->shards, *u );
+						state vector<int> execIdxVec;
+						state std::deque< Standalone<VerUpdateRef> >::iterator u = updatesToSplit.begin();
+						for(; u != updatesToSplit.end(); ++u) {
+							ASSERT(execIdxVec.size() == 0);
+							splitMutations(data, data->shards, *u, execIdxVec);
+							for (auto execIdx : execIdxVec) {
+								wait(snapHelper(data, u->mutations[execIdx], u->version));
+							}
+							execIdxVec.clear();
+						}
```

Comment:
Sorry, I'm confused, but it's likely just since I don't actually know the internal flow of the storage server.  I don't see anything in splitMutations that's writing to disk, so why do you need to snapshot here?

Reply:
splitMutation ends up calling the function that adds these mutations to the mutation log (splitMutation->addMutation->addMutationToMutationLog). From looking through the code, I felt this was the best place to intercept the call and take the snapshot and avoid having the exec mutation make it to disk.